clusterloader2/pkg/measurement/common/pod_startup_latency.go

```diff
@@ -0,0 +1,293 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement"
+	measurementutil "k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+)
+
+const (
+	podStartupLatencyMetricName = "PodStartupLatency"
+	informerSyncTimeout         = time.Minute
+)
+
+func init() {
+	measurement.Register(podStartupLatencyMetricName, createPodStartupLatencyMeasurement)
+}
+
+func createPodStartupLatencyMeasurement() measurement.Measurement {
+	return &podStartupLatencyMeasurement{
+		createTimes:   make(map[string]metav1.Time),
+		scheduleTimes: make(map[string]metav1.Time),
+		runTimes:      make(map[string]metav1.Time),
+		watchTimes:    make(map[string]metav1.Time),
+	}
+}
+
+type podStartupLatencyMeasurement struct {
+	namespace     string
+	labelSelector string
+	fieldSelector string
+	informer      cache.SharedInformer
+	stopCh        chan struct{}
+	mutex         sync.Mutex
+	createTimes   map[string]metav1.Time
+	scheduleTimes map[string]metav1.Time
+	runTimes      map[string]metav1.Time
+	watchTimes    map[string]metav1.Time
+}
+
+// Execute supports two actions:
+// - start - Starts to observe pods and pods events.
+// - gather - Gathers and prints current pod latency data.
+func (p *podStartupLatencyMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	var summaries []measurement.Summary
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return summaries, err
+	}
+
+	switch action {
+	case "start":
+		p.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return summaries, err
+		}
+		p.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		p.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		return summaries, p.start(config.ClientSet)
+	case "gather":
+		return p.gather(config.ClientSet)
+	default:
+		return summaries, fmt.Errorf("unknown action %v", action)
+	}
+
+}
+
+func (p *podStartupLatencyMeasurement) start(c clientset.Interface) error {
+	glog.Infof("Starting pod startup latency measurement...")
```

Comment:
You should check if it wasn't already started before.

Reply:
Done.